builtin/providers/aws/config.go

```diff
@@ -316,3 +325,28 @@ func (c *Config) ValidateAccountId(iamconn *iam.IAM) error {
 
 	return nil
 }
+
+// This function is responsible for reading credentials from the
+// environment in the case that they're not explicitly specified
+// in the Terraform configuration.
+func getCreds(key, secret, token string) *awsCredentials.Credentials {
+	// build a chain provider, lazy-evaulated by aws-sdk
+	providers := []awsCredentials.Provider{
+		&awsCredentials.StaticProvider{Value: awsCredentials.Value{
+			AccessKeyID:     key,
+			SecretAccessKey: secret,
+			SessionToken:    token,
+		}},
+		&awsCredentials.EnvProvider{},
+		&awsCredentials.SharedCredentialsProvider{},
+	}
+
+	// We only look in the EC2 metadata API if we can connect
+	// to the metadata service within a reasonable amount of time
+	conn, err := net.DialTimeout("tcp", "169.254.169.254:80", 100*time.Millisecond)
```

Comment:
I'll mock this out with `httptest` server, pulling the metadata url from an ENV var for testing. [Nomad][1] does [this][2]

[1]: https://github.com/hashicorp/nomad/blob/master/client/fingerprint/env_aws.go#L20
[2]: https://github.com/hashicorp/nomad/blob/master/client/fingerprint/env_aws_test.go#L43